doc/development/plugins.rst

```diff
@@ -1,397 +1,318 @@
+(array([0, 0, 0, 0, 0]), array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))
+
+The :class:`~.measurements.Shots` class describes the shots. Users can optionally specify a shot vector, or
+different numbers of shots to use when calculating the final expecation value.
+
+>>> tape0 = qml.tape.QuantumScript([], [qml.expval(qml.PauliX(0))], shots=(5, 500, 1000))
+>>> tape0.shots.shot_vector
+(ShotCopies(5 shots x 1),
+ ShotCopies(500 shots x 1),
+ ShotCopies(1000 shots x 1))
+>>> list(tape0.shots)
+[5, 500, 1000]
+>>> list(tape0.shots.bins())
+[(0, 5), (5, 505), (505, 1505)]
+>>> dev.execute(tape0)
+(0.2, -0.052, -0.014)
+
+The first number ``0.2`` is calculated with 5 shots, the second ``-0.052`` is calculated with 500 shots, and
+``-0.014`` is calculated with 1000 shots.  All 1,505 shots can be requested together in one batch, but the post
+processing into the expecation value is done with shots ``0:5``, ``5:505``, and ``505:1505`` respectively.
+
+Wires
+-----
+
+Devices can now either:
+
+1) Strictly use wires provided by the user on initialization ``device(name, wires=wires)``
+2) Infer the number and ordering of wires provided by the submitted circuit.
+3) Strictly require specific wire labels
+
+Option 2 allows workflows to change the number and labeling of wires over time, but sometimes users want
+to enfore a wire convention and labels. If a user does provide wires, the :meth:`~.devices.Device.preprocess` should
+validate that submitted circuits only have wires in the requested range.
+
+>>> dev = qml.device('default.qubit', wires=1)
+>>> circuit = qml.tape.QuantumScript([qml.CNOT((0,1))], [qml.state()])
+>>> dev.preprocess()[0]((circuit, ))
+WireError: Cannot run circuit(s) of default.qubit as they contain wires not found on the device.
+
+PennyLane wires can be any hashable object, where wire labels are distinguished by their equality and hash.
+If working with successive integers (``0``, ``1``, ``2``, ...) is preferred internally,
+the :meth:`~.QuantumScript.map_to_standard_wires` method can be used inside of 
+the :meth:`~.devices.Device.execute` method.  The :class:`~.map_wires` transform can also 
+map the wires of the submitted circuit to internal labels.
+
+Sometimes hardware qubit labels cannot be arbitrarily mapped without a changing in behaviour.
+Connectivity, noise, performance, and
+other constraints can make it so that operations on qubit one cannot be arbitrarily exchanged with the same operation
+of qubit two.  In such a situation, the device can hard code a list of the only acceptable wire labels. In such a case, it
+will be on the user to deliberately map wires if they wish such a thing to occur.
+
+>>> qml.device('my_hardware').wires
+<Wires = [0, 1, 2, 3]>
+>>> qml.device('my_hardware', wires=(10, 11, 12, 13))
+DeviceError: Device my_hardware cannot internally map wires, as the labels 0, 1, 2, 3 indicate unique qubits
+
+
+Preprocessing
+-------------
 
-    apply
+The preprocessing method has two main responsibilities:
```

Comment:
Maybe to put it into perspective, we can briefly explain that a PL circuit is first transformed by the transform program returned by `preprocess`, and then the transformed circuits are sent to `execute` to be executed. This will help the developer understand the purpose of `preprocess` better.